README
NAME
    RT-Extension-ListAllEmailRecipients - Determine all recipients for a
    notification and provide emails via template variables

DESCRIPTION
    ListAllEmailRecipients does a dry run of all scrips configured for a
    notification to determine the full list of email receipients. This list
    is then made available to templates when the actual notification scrips
    are subsequently run to send email.

RT VERSION
    Works with RT 4.4

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        Add these lines, both are required:

            Plugin('RT::Extension::ListAllEmailRecipients');
            Set(@MailPlugins, qw(Auth::MailFrom Action::ListAllEmailRecipients));

    Clear your mason cache
            rm -rf /opt/rt4/var/mason_data/obj

    Restart your webserver
    Modify your templates as described below.

USAGE
    ListEmailReceipients adds the following template variables containing a
    list of all recipients for that type for the current transaction
    (comment, correspond, etc.) across all enabled notification scrips.

        $NotificationRecipientsTo
        $NotificationRecipientsCc
        $NotificationRecipientsBcc

    To include these in an outgoing email, like the Admin email to AdminCcs
    on a ticket, add something like the following to the appropriate
    template:

        <p>Email was sent to the following addresses:</p>

        <p>To: {$NotificationRecipientsTo} </p>
        <p>Cc: {$NotificationRecipientsCc} </p>
        <p>Bcc: {$NotificationRecipientsBcc} </p>

  Ticket and Transaction IDs
    This extension generates the recipient lists by doing a trial run of the
    incoming action (create, comment, or reply). In doing so, it doesn't
    actually make any changes, but it does increment the ids in the ticket
    and transaction tables and possibly others depending on your scrips.
    This shouldn't matter for most users since ticket ids are arbitrary, but
    some users depend on ticket ids for various reasons. If you have
    processes that depend on specific ticket ids, be aware that using this
    extension will create gaps between ticket and other ids in your
    database.

AUTHOR
    Best Practical Solutions, LLC <modules@bestpractical.com>

BUGS
    Contact Best Practical Solutions at contact@bestpractical.com with
    questions about this extension.

LICENSE AND COPYRIGHT
    This software is Copyright (c) 2016 by Best Practical Solutions, LLC

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
Revision history for RT-Extension-ListAllEmailRecipients

0.03 2017-01-31
 - Fix issues with TransactionBatch scrips

0.02 2016-12-21
 - Populate recipients on create as well as update
 - Add compatibility with RT::Extension::OneTimeTo
 - Clear recipient values to avoid possible retention of values from previous notification

0.01
 - Initial version


lib/RT/Extension/ListAllEmailRecipients/Test.pm
use strict;
use warnings;

### after: use lib qw(@RT_LIB_PATH@);
use lib qw(/opt/rt4/local/lib /opt/rt4/lib);

package RT::Extension::ListAllEmailRecipients::Test;

our @ISA;
BEGIN {
    local $@;
    eval { require RT::Test; 1 } or do {
        require Test::More;
        Test::More::BAIL_OUT(
            "requires 3.8 to run tests. Error:\n$@\n"
            ."You may need to set PERL5LIB=/path/to/rt/lib"
        );
    };
    push @ISA, 'RT::Test';
}

sub import {
    my $class = shift;
    my %args  = @_;

    $args{'requires'} ||= [];
    if ( $args{'testing'} ) {
        unshift @{ $args{'requires'} }, 'RT::Extension::ListAllEmailRecipients';
    } else {
        $args{'testing'} = 'RT::Extension::ListAllEmailRecipients';
    }

    $class->SUPER::import( %args );
    $class->export_to_level(1);

    require RT::Extension::ListAllEmailRecipients;
}


1;


lib/RT/Template_Vendor.pm
package RT::Template;

use strict;
use warnings;
no warnings qw(redefine);

# Copied from RT's core RT::Template::Parse to add a variable to the template.
sub Parse {
    my $self = shift;
    my ($rv, $msg);

    my @more_template_args = $self->AddTemplateArgs(@_);

    if (not $self->IsEmpty and $self->Content =~ m{^Content-Type:\s+text/html\b}im) {
        local $RT::Transaction::PreferredContentType = 'text/html';
        ($rv, $msg) = $self->_Parse(@_, @more_template_args);
    }
    else {
        ($rv, $msg) = $self->_Parse(@_, @more_template_args);
    }

    return ($rv, $msg) unless $rv;

    my $mime_type   = $self->MIMEObj->mime_type;
    if (defined $mime_type and $mime_type eq 'text/html') {
        $self->_DowngradeFromHTML(@_, @more_template_args);
    }

    return ($rv, $msg);
}

sub AddTemplateArgs {
    my $self = shift;
    my @more_template_args;
    push @more_template_args,
        'NotificationRecipientsTo' => $RT::Extension::ListAllEmailRecipients::NotificationRecipientsTo,
        'NotificationRecipientsCc' => $RT::Extension::ListAllEmailRecipients::NotificationRecipientsCc,
        'NotificationRecipientsBcc' => $RT::Extension::ListAllEmailRecipients::NotificationRecipientsBcc;

    return @more_template_args;
}

1;


lib/RT/Action/Notify_Vendor.pm
package RT::Action::Notify;

use strict;
use warnings;
no warnings 'redefine';

=head1 Prepare

After the default Notify.pm SetRecipients and the SendEmail
prepare, the MIME object is created and populated. Use it to
pull email headers and add to the message body the list of
email addresses who will receive the email.

=cut

sub Prepare {
    my $self = shift;
    $self->SetRecipients();
    my $return = $self->SUPER::Prepare();

    # Now that the template is prepared, check for recipients and
    # add them to the template if needed.
    $self->AddRecipientsToEmail();
    return $return;
}

=head1 AddRecipientsToEmail

Inspect the created MIME object, pull out To and Cc headers,
and add the list of emails to the body of the email, replacing
the RT-INSERT-RECIPIENTS placeholder.

=cut

sub AddRecipientsToEmail {
    my $self = shift;
    my $ent = $self->TemplateObj->MIMEObj;
    my @parts;

    # Could be one plain text part or multiple parts with plain
    # text and html. Handle both.
    if ( $ent->mime_type and $ent->mime_type eq 'text/plain' ){
        push @parts, $ent;
    }
    else{
        ExtractParts({ Parts => \@parts, Entity => $ent });
    }

    foreach my $part (@parts){
        $RT::Logger->info("Working on email part with type: " . $part->mime_type)
            if $part->mime_type;
        next unless $part->mime_type eq 'text/plain'
            or $part->mime_type eq 'text/html';

        # Read the (unencoded) body data
        my @original;
        if ( my $io = $part->open("r") ) {
            while ( my $line = $io->getline ) {
                push @original, $line if defined $line;
            }
            $io->close;
        }

        # Write back the message, replacing the recipient
        # placeholder with the actual recipient list.
        if ( my $io = $part->open("w") ) {
            foreach my $line (@original) {

                if ( $line =~ /RT-INSERT-RECIPIENTS/ ){
                    # Replace with list of recipients and write to email
                    my $emails = '';
                    $emails = join ', ', $self->AddressesFromHeader('To'),
                        $self->AddressesFromHeader('Cc');

                    $RT::Logger->info("Replacing RT-INSERT-RECIPIENTS with $emails");
                    $line =~ s/(RT\-INSERT\-RECIPIENTS)/$emails/;
                }

                $io->print($line);
            }
            $io->close;
        }
    }

    return;
}

sub ExtractParts {
    my $args_ref = shift;
    my $parts_ref = $args_ref->{'Parts'};
    my $ent = $args_ref->{'Entity'};

    # Drill down into multipart message until there are no more parts
    foreach my $part ($ent->parts){
        if ( $part->parts ){
            ExtractParts({ Parts => $parts_ref, Entity => $part });
        }
        else{
            push @{$parts_ref}, $part
            if ($part->mime_type eq 'text/plain' or
                $part->mime_type eq 'text/html');
        }
    }
    return;
}

1;


lib/RT/Interface/Email/Action/Defaults_Vendor.pm
package RT::Interface::Email::Action::Defaults;

use strict;
use warnings;
no warnings qw(redefine);

sub _HandleEither {
    my %args = (
        Action      => undef,
        Message     => undef,
        Subject     => undef,
        Ticket      => undef,
        TicketId    => undef,
        Queue       => undef,
        @_,
    );

    return _HandleCreate(@_) unless $args{TicketId};

    unless ( $args{Ticket}->Id ) {
        MailError(
            Subject     => "Message not recorded: $args{Subject}",
            Explanation => "Could not find a ticket with id " . $args{TicketId},
            FAILURE     => 1,
        );
    }

    RT::Extension::ListAllEmailRecipients::FindNotificationRecipients(
        @_,
        UpdateInterface => 'Email');

    my $action = ucfirst $args{Action};
    my ( $status, $msg ) = $args{Ticket}->$action( MIMEObj => $args{Message} );
    return if $status;
}

1;


lib/RT/Extension/ListAllEmailRecipients.pm
use strict;
use warnings;
package RT::Extension::ListAllEmailRecipients;

our $VERSION = '0.03';

=head1 NAME

RT-Extension-ListAllEmailRecipients - Determine all recipients for a
notification and provide emails via template variables

=head1 DESCRIPTION

ListAllEmailRecipients does a dry run of all scrips configured for a
notification to determine the full list of email receipients. This
list is then made available to templates when the actual notification
scrips are subsequently run to send email.

=head1 RT VERSION

Works with RT 4.4

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt4/etc/RT_SiteConfig.pm>

Add these lines, both are required:

    Plugin('RT::Extension::ListAllEmailRecipients');
    Set(@MailPlugins, qw(Auth::MailFrom Action::ListAllEmailRecipients));

=item Clear your mason cache

    rm -rf /opt/rt4/var/mason_data/obj

=item Restart your webserver

=item Modify your templates as described below.

=back

=head1 USAGE

ListEmailReceipients adds the following template variables containing
a list of all recipients for that type for the current transaction
(comment, correspond, etc.) across all enabled notification scrips.

    $NotificationRecipientsTo
    $NotificationRecipientsCc
    $NotificationRecipientsBcc

To include these in an outgoing email, like the Admin email to AdminCcs
on a ticket, add something like the following to the appropriate template:

    <p>Email was sent to the following addresses:</p>

    <p>To: {$NotificationRecipientsTo} </p>
    <p>Cc: {$NotificationRecipientsCc} </p>
    <p>Bcc: {$NotificationRecipientsBcc} </p>

=head2 Ticket and Transaction IDs

This extension generates the recipient lists by doing a trial run of the
incoming action (create, comment, or reply). In doing so, it doesn't actually
make any changes, but it does increment the ids in the ticket and transaction
tables and possibly others depending on your scrips. This shouldn't matter for
most users since ticket ids are arbitrary, but some users depend on ticket ids
for various reasons. If you have processes that depend on specific ticket ids,
be aware that using this extension will create gaps between ticket and other
ids in your database.

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

Contact Best Practical Solutions at contact@bestpractical.com with
questions about this extension.

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2016 by Best Practical Solutions, LLC

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

our $NotificationRecipientsTo;
our $NotificationRecipientsCc;
our $NotificationRecipientsBcc;

sub FindNotificationRecipients {
    my %args = (
        TicketId => undef,
        @_
    );

    $NotificationRecipientsTo = '';
    $NotificationRecipientsCc = '';
    $NotificationRecipientsBcc = '';

    my $recipients_ref = ProcessScripDryRun(%args);
    $NotificationRecipientsTo = join ', ', @{$recipients_ref->{'To'}} if $recipients_ref->{'To'};
    $NotificationRecipientsCc = join ', ', @{$recipients_ref->{'Cc'}} if $recipients_ref->{'Cc'};
    $NotificationRecipientsBcc = join ', ', @{$recipients_ref->{'Bcc'}} if $recipients_ref->{'Bcc'};

    return;
}

# Code copied from ShowSimplifiedRecipients and modified

sub ProcessScripDryRun {
    my %args = (
        UpdateInterface => 'Web',
        @_);

    RT::Logger->debug("Starting dry run to gather recipients");

    my @dryrun;

    if ( $args{'UpdateInterface'} eq 'EmailCreate' ) {
        # load separate ticket obj for dry-run
        my $TicketObj = RT::Ticket->new($args{'Ticket'}->CurrentUser);
        @dryrun = $TicketObj->DryRun(
            sub {
                $TicketObj->Create(
                    Queue     => $args{'Queue'},
                    Subject   => $args{'Subject'},
                    Requestor => $args{'Requestor'},
                    Cc        => $args{'Cc'},
                    MIMEObj   => $args{'MIMEObj'},
                );
            }
        );
    }
    elsif ( $args{'UpdateInterface'} eq 'Email' ){
        # load separate ticket obj for dry-run
        my $TicketObj = RT::Ticket->new($args{'CurrentUser'});
        my ($ret, $msg) = $TicketObj->Load($args{Ticket}->Id);

        unless ( $ret ){
            RT::Logger->error("Unable to load ticket " . $args{'Ticket'}->Id . ". Skipping dry run.");
            return;
        }

        my $action = ucfirst $args{Action};
        @dryrun = $TicketObj->DryRun(
            sub {
                my ( $status, $msg ) = $TicketObj->$action( MIMEObj => $args{Message} );
            }
        );
    }
    elsif ( $args{'id'} and $args{'id'} eq 'new' ){
        return unless $args{'TicketObj'};
        # load separate ticket obj for dry-run
        my $TicketObj = RT::Ticket->new($args{CurrentUser});

        @dryrun = $TicketObj->DryRun(
            sub {
                local $args{UpdateContent} ||= "Content";
                HTML::Mason::Commands::CreateTicket( %args );
            }
        );

    }
    else{
        # load separate ticket obj for dry-run
        my $TicketObj = RT::Ticket->new($args{'CurrentUser'});
        my ($ret, $msg) = $TicketObj->Load($args{'TicketId'});

        unless ( $ret ){
            RT::Logger->error("Unable to load ticket " . $args{'TicketId'} . ". Skipping dry run.");
            return;
        }

        @dryrun = $TicketObj->DryRun(
            sub {
                local $args{UpdateContent} ||= "Content";
                HTML::Mason::Commands::ProcessUpdateMessage(ARGSRef  => \%args, TicketObj => $TicketObj, KeepAttachments => 1 );
                HTML::Mason::Commands::ProcessTicketWatchers(ARGSRef => \%args, TicketObj => $TicketObj );
                HTML::Mason::Commands::ProcessTicketBasics(  ARGSRef => \%args, TicketObj => $TicketObj );
                HTML::Mason::Commands::ProcessTicketLinks(   ARGSRef => \%args, TicketObj => $TicketObj );
                HTML::Mason::Commands::ProcessTicketDates(   ARGSRef => \%args, TicketObj => $TicketObj );
                HTML::Mason::Commands::ProcessObjectCustomFieldUpdates(ARGSRef => \%args, TicketObj => $TicketObj );
                HTML::Mason::Commands::ProcessTicketReminders( ARGSRef => \%args, TicketObj => $TicketObj );
            }
        );
    }
    return unless @dryrun;

    my %headers = (To => {}, Cc => {}, Bcc => {});
    my %no_squelch = (To => {}, Cc => {}, Bcc => {});
    my @scrips = map {@{$_->Scrips->Prepared}} @dryrun;
    if (@scrips) {
        for my $scrip (grep $_->ActionObj->Action->isa('RT::Action::SendEmail'), @scrips) {
            my $action = $scrip->ActionObj->Action;
            for my $type (qw(To Cc Bcc)) {
                for my $addr ($action->$type()) {
                    if (grep {$addr->address eq $_} @{$action->{NoSquelch}{$type} || []}) {
                        $no_squelch{$type}{$addr->address} = $addr;
                    } else {
                        $headers{$type}{$addr->address} = $addr;
                    }
                }
            }
        }
    }

    my %squelched = HTML::Mason::Commands::ProcessTransactionSquelching( \%args );
    my $squelched_config = !( RT->Config->Get('SquelchedRecipients', $args{'CurrentUser'}) );

    # Unpack the addresses found and save
    my %recipients;
    for my $type (qw(To Cc Bcc)) {
        next unless keys %{$headers{$type}} or keys %{$no_squelch{$type}};

        for my $addr (sort {$a->address cmp $b->address} values %{$headers{$type}}) {
            push @{$recipients{$type}}, $addr->address;
        }
    }

    # One-time Ccs and Bccs are processed through email headers, so they won't be picked up
    # by running the action To, Cc, Bcc methods above. If passed, add them here.
    foreach my $one_time ( qw(To Cc Bcc) ){
        my @addresses = Email::Address->parse($args{"Update" . $one_time}) if $args{"Update" . $one_time};
        push @{$recipients{$one_time}}, @addresses;
    }

    RT::Logger->debug("Found some recipients") if %recipients;

    return \%recipients;
}

1;
